File: backend/src/routes/imports.ts
import { Router } from 'express';
import multer from 'multer';
import path from 'node:path';
import { UPLOAD_DIR } from '../config.js';
import { db } from '../db/index.js';
import {
  parseUpload,
  parsePdf,
  getParsed,
  applyMapping,
  transformDegiro,
  DEFAULT_ACTION_MAP,
} from '../services/importer.js';
import { resolveInstrument, insertTransaction, makeDedupeKey } from '../services/repo.js';
import type { ImportMapping, ImportPreviewRow } from '@decisionguru/shared';

const upload = multer({ dest: UPLOAD_DIR, limits: { fileSize: 25 * 1024 * 1024 } });
export const importsRouter = Router();

importsRouter.post('/upload', upload.single('file'), async (req, res) => {
  if (!req.file) return res.status(400).json({ error: 'No file uploaded' });
  const ext = path.extname(req.file.originalname).toLowerCase();
  try {
    const parsed = ext === '.pdf' ? await parsePdf(req.file.path, req.file.originalname) : parseUpload(req.file.path, req.file.originalname);
    res.json({ ...parsed, defaultActionMap: DEFAULT_ACTION_MAP });
  } catch (err) {
    res.status(422).json({ error: `Could not parse file: ${(err as Error).message}` });
  }
});

importsRouter.get('/file/:fileId', (req, res) => {
  const parsed = getParsed(req.params.fileId);
  if (!parsed) return res.status(404).json({ error: 'File not found (re-upload)' });
  res.json(parsed);
});

/** Preview rows for either a broker preset or a manual column mapping. */
function buildPreviewRows(mapping: ImportMapping): ImportPreviewRow[] {
  if (mapping.broker === 'degiro') return transformDegiro(mapping.fileId, mapping.sheetName);
  return applyMapping(mapping);
}

importsRouter.post('/preview', (req, res) => {
  const mapping = req.body as ImportMapping;
  if (!mapping?.fileId) return res.status(400).json({ error: 'fileId required' });
  if (!getParsed(mapping.fileId)) return res.status(404).json({ error: 'File not found (re-upload)' });
  const rows = buildPreviewRows(mapping);
  const okCount = rows.filter((r) => r.ok).length;
  const corporateActions = rows.filter((r) => r.category === 'corporate_action').length;
  const trades = rows.filter((r) => r.ok && r.category !== 'corporate_action').length;
  res.json({ rows, okCount, total: rows.length, trades, corporateActions });
});

importsRouter.post('/commit', async (req, res) => {
  const mapping = req.body as ImportMapping;
  if (!mapping?.fileId) return res.status(400).json({ error: 'fileId required' });
  if (!getParsed(mapping.fileId)) return res.status(404).json({ error: 'File not found (re-upload)' });
  const rows = buildPreviewRows(mapping).filter((r) => r.ok);

  let imported = 0;
  let skipped = 0;
  let corporateActions = 0;
  const instrumentCache = new Map<string, number>();
  const touched = new Set<number>();

  for (const row of rows) {
    const t = row.tx;
    const key = (t.symbol || t.isin || t.name || '').toUpperCase();
    let instrumentId = instrumentCache.get(key);
    if (instrumentId === undefined) {
      const inst = await resolveInstrument({ symbol: t.symbol, isin: t.isin, name: t.name });
      instrumentId = inst.id;
      instrumentCache.set(key, instrumentId);
    }
    const dedupeKey = (t as { dedupeKey?: string }).dedupeKey ?? makeDedupeKey({ ...t, instrumentId });
    const before = db.prepare('SELECT COUNT(*) AS c FROM transactions').get() as { c: number };
    insertTransaction({
      instrumentId,
      action: t.action!,
      date: t.date!,
      quantity: t.quantity ?? 0,
      unitPrice: t.unitPrice ?? 0,
      fees: t.fees ?? 0,
      currency: t.currency,
      grossAmount: t.grossAmount ?? null,
      netAmount: t.netAmount ?? null,
      withholding: t.withholding ?? null,
      category: row.category ?? 'trade',
      note: t.note ?? null,
      source: `import:${mapping.broker ?? mapping.fileId}`,
      dedupeKey,
    });
    const after = db.prepare('SELECT COUNT(*) AS c FROM transactions').get() as { c: number };
    if (after.c > before.c) {
      imported += 1;
      if (row.category === 'corporate_action') corporateActions += 1;
      touched.add(instrumentId);
    } else {
      skipped += 1;
    }
  }

  res.json({ imported, skipped, corporateActions, instruments: [...touched] });
});

// ---- presets ------------------------------------------------------------

importsRouter.get('/presets', (_req, res) => {
  const rows = db.prepare('SELECT * FROM import_presets ORDER BY name').all() as any[];
  res.json(rows.map((r) => ({ ...r, mapping: JSON.parse(r.mapping) })));
});

importsRouter.post('/presets', (req, res) => {
  const { name, mapping } = req.body ?? {};
  if (!name || !mapping) return res.status(400).json({ error: 'name and mapping required' });
  db.prepare(
    'INSERT INTO import_presets (name, mapping) VALUES (?, ?) ON CONFLICT(name) DO UPDATE SET mapping = excluded.mapping',
  ).run(name, JSON.stringify(mapping));
  res.json({ ok: true });
});

importsRouter.delete('/presets/:id', (req, res) => {
  db.prepare('DELETE FROM import_presets WHERE id = ?').run(Number(req.params.id));
  res.json({ ok: true });
});

File: backend/app/routers/imports.py
from __future__ import annotations

import json
import secrets
from pathlib import Path

from fastapi import APIRouter, Request, UploadFile
from fastapi.concurrency import run_in_threadpool

from ..core import db
from ..core.config import UPLOAD_DIR
from ..core.errors import ApiError
from ..services import repo
from ..services.importer import (
    DEFAULT_ACTION_MAP,
    apply_mapping,
    get_parsed,
    parse_pdf,
    parse_upload,
    transform_degiro,
)

router = APIRouter()


@router.post("/upload")
async def upload(file: UploadFile | None = None) -> dict:
    if file is None:
        raise ApiError("No file uploaded", 400)
    ext = Path(file.filename or "").suffix.lower()
    dest = UPLOAD_DIR / secrets.token_hex(16)
    dest.write_bytes(await file.read())
    try:
        if ext == ".pdf":
            parsed = await run_in_threadpool(parse_pdf, str(dest), file.filename)
        else:
            parsed = await run_in_threadpool(parse_upload, str(dest), file.filename)
        return {**parsed, "defaultActionMap": DEFAULT_ACTION_MAP}
    except Exception as exc:  # noqa: BLE001
        raise ApiError(f"Could not parse file: {exc}", 422)


@router.get("/file/{file_id}")
async def get_file(file_id: str) -> dict:
    parsed = get_parsed(file_id)
    if not parsed:
        raise ApiError("File not found (re-upload)", 404)
    return parsed


def _build_preview_rows(mapping: dict) -> list[dict]:
    if mapping.get("broker") == "degiro":
        return transform_degiro(mapping["fileId"], mapping.get("sheetName"))
    return apply_mapping(mapping)


@router.post("/preview")
async def preview(request: Request) -> dict:
    mapping = await request.json() or {}
    if not mapping.get("fileId"):
        raise ApiError("fileId required", 400)
    if not get_parsed(mapping["fileId"]):
        raise ApiError("File not found (re-upload)", 404)
    rows = await run_in_threadpool(_build_preview_rows, mapping)
    ok_count = sum(1 for r in rows if r["ok"])
    corporate_actions = sum(1 for r in rows if r.get("category") == "corporate_action")
    trades = sum(1 for r in rows if r["ok"] and r.get("category") != "corporate_action")
    return {"rows": rows, "okCount": ok_count, "total": len(rows),
            "trades": trades, "corporateActions": corporate_actions}


@router.post("/commit")
async def commit(request: Request) -> dict:
    mapping = await request.json() or {}
    if not mapping.get("fileId"):
        raise ApiError("fileId required", 400)
    if not get_parsed(mapping["fileId"]):
        raise ApiError("File not found (re-upload)", 404)

    def _work() -> dict:
        rows = [r for r in _build_preview_rows(mapping) if r["ok"]]
        imported = 0
        skipped = 0
        corporate_actions = 0
        instrument_cache: dict[str, int] = {}
        touched: set[int] = set()

        for row in rows:
            t = row["tx"]
            key = (t.get("symbol") or t.get("isin") or t.get("name") or "").upper()
            instrument_id = instrument_cache.get(key)
            if instrument_id is None:
                inst = repo.resolve_instrument(
                    {"symbol": t.get("symbol"), "isin": t.get("isin"), "name": t.get("name")}
                )
                instrument_id = inst["id"]
                instrument_cache[key] = instrument_id
            dedupe_key = t.get("dedupeKey") or repo.make_dedupe_key({**t, "instrumentId": instrument_id})
            before = repo.transactions_count()
            repo.insert_transaction({
                "instrumentId": instrument_id,
                "action": t.get("action"),
                "date": t.get("date"),
                "quantity": t.get("quantity") or 0,
                "unitPrice": t.get("unitPrice") or 0,
                "fees": t.get("fees") or 0,
                "currency": t.get("currency"),
                "grossAmount": t.get("grossAmount"),
                "netAmount": t.get("netAmount"),
                "withholding": t.get("withholding"),
                "category": row.get("category") or "trade",
                "note": t.get("note"),
                "source": f"import:{mapping.get('broker') or mapping['fileId']}",
                "dedupeKey": dedupe_key,
            })
            after = repo.transactions_count()
            if after > before:
                imported += 1
                if row.get("category") == "corporate_action":
                    corporate_actions += 1
                touched.add(instrument_id)
            else:
                skipped += 1

        return {"imported": imported, "skipped": skipped,
                "corporateActions": corporate_actions, "instruments": list(touched)}

    return await run_in_threadpool(_work)


# ---- presets ----------------------------------------------------------------

@router.get("/presets")
async def list_presets() -> list[dict]:
    rows = db.q("SELECT * FROM import_presets ORDER BY name").all()
    return [{**dict(r), "mapping": json.loads(r["mapping"])} for r in rows]


@router.post("/presets")
async def save_preset(request: Request) -> dict:
    body = await request.json() or {}
    if not body.get("name") or not body.get("mapping"):
        raise ApiError("name and mapping required", 400)
    db.execute(
        "INSERT INTO import_presets (name, mapping) VALUES (?, ?) "
        "ON CONFLICT(name) DO UPDATE SET mapping = excluded.mapping",
        (body["name"], json.dumps(body["mapping"])),
    )
    return {"ok": True}


@router.delete("/presets/{preset_id}")
async def delete_preset(preset_id: int) -> dict:
    db.execute("DELETE FROM import_presets WHERE id = ?", (preset_id,))
    return {"ok": True}
